lib/streamlit/runtime/caching/cache_utils.py

```diff
@@ -417,9 +543,31 @@ def save_element_message(
         """
         id_to_save = self.select_dg_to_save(invoked_dg_id, used_dg_id)
         for msgs in self._cached_message_stack:
-            msgs.append(
-                ElementMsgData(delta_type, element_proto, id_to_save, returned_dg_id)
-            )
+            try:
+                wid = element_proto.id  # type: ignore
+                # TODO replace `Message` with a more precise type
+                if not self._registered_metadata:
+                    _LOGGER.warning(
+                        "Trying to save widget message that wasn't registered. This should not happen."
+                    )
+                    raise AttributeError
```

Comment:
If streamlit is running, it should never happen, and indicates something is broken. But, if streamlit isn't running the script, widget registration never happens, yet we can still end up in that caching code. The current behavior here was a hack to get integration tests working. Though now that I'm thinking about it again, I could check if the runtime exists, so it can silently continue or explode, as appropriate.